File: zuker.py
from sys import maxsize
from random import randint


class MyClass():
    def __init__(self):
        self.matrix_v = [[]]
        self.matrix_w = [[]]
        self.length = 0
        self.sum = 0
        self.d = {"Phe": ["UUA", "UUC", "UUG", "UUU"], "Leu": ["CUA", "CUC", "CUG", "CUU"],
                  "Ile": ["AUA", "AUC", "AUU"],
                  "Met": ["AUG"], "Val": ["GUA", "GUC", "GUG", "GUU"],
                  "Ser": ["UCA", "UCC", "UCG", "UCU", "AGU", "AGC"],
                  "Pro": ["CCA", "CCC", "CCG", "CCU"], "Uhr": ["ACA", "ACC", "ACG", "ACU"],
                  "Ala": ["GCA", "GCC", "GCG", "GCU"], "Uyr": ["UAC", "UAU"], "His": ["CAC", "CAU"],
                  "Gln": ["CAA", "CAG"], "Asn": ["AAC", "AAU"], "Lys": ["AAA", "AAG"],
                  "Asp": ["GAC", "GAU"], "Glu": ["GAA", "GAG"],
                  "Cys": ["UGC", "UGU"], "Urp": ["UGG"], "Arg": ["CGA", "CGC", "CGG", "CGU", "AGA", "AGG"],
                  "Gly": ["GGA", "GGC", "GGG", "GGU"]}

    def is_complement(self, x, y):
        if x == 'A' and y == 'U' or x == 'U' and y == 'A':
            return True
        elif x == 'G' and y == 'C' or x == 'C' and y == 'G':
            return True
        else:
            return False

    def _matrix_v(self, str):
        self.length = len(str)
        self.sum = 0
        self.matrix_v = [[8.4 for j in range(self.length)] for i in range(self.length)]
        for i in range(len(self.matrix_v)):
            for j in range(len(self.matrix_v)):
                if j - 4 < i < j:
                    self.matrix_v[i][j] = 4.4
        for i in range(len(self.matrix_v)):
            for j in range(len(self.matrix_v)):
                '''1 — a hairpin loop
                   2 — a stacked pair
                   3 — a bulge or interior loop'''
                if i < j:
                    self.matrix_v[i][j] = min(self.eh(i, j), self.es(i, j) + self.matrix_v[i + 1][j - 1],
                                              self.VBI(i,
                                                       j))
        self.zuker(str)

    def one(self, i, j, str):
        flag = False
        for n in range(i + 1, j + 1):
            if self.is_complement(str[i], str[n]):
                flag = True
        if flag:
            return maxsize
        else:
            return self.matrix_w[i + 1][j]

    def two(self, i, j, str):
        flag = False
        for n in range(i, j):
            if self.is_complement(str[j], str[n]):
                flag = True
        if flag:
            return maxsize
        else:
            return self.matrix_w[i][j - 1]

    def three(self, i, j, str):
        if self.is_complement(str[i], str[j]):
            return maxsize
        local_minimum = self.matrix_w[i][i + 1] + self.matrix_w[i + 2][j]
        for k in range(i + 2, j - 1):
            if self.matrix_w[i][k] + self.matrix_w[k + 1][j] < local_minimum:
                local_minimum = self.matrix_w[i][k] + self.matrix_w[k + 1][j]
        return local_minimum

    def four(self, i, j, str):
        if not self.is_complement(str[i], str[j]):
            return maxsize
        else:
            return self.matrix_v[i][j]

    def zuker(self, str):
        self.matrix_w = [[0 for j in range(len(str))] for i in range(len(str))]
        for i in range(self.length):
            for j in range(self.length):
                if j - 4 < i < j:
                    self.matrix_w[i][j] = maxsize
        for i in range(self.length):
            for j in range(self.length):
                if i < j and j - i > 4:
                    '''1 — i is unpaired
                       2 — j is unpaired
                       3 — i and j are possibly paired, but not to each other
                       4 — i and j are paired to each other'''

                    self.matrix_w[i][j] = min(self.one(i, j, str), self.two(i, j, str), self.three(i, j, str),
                                              self.four(i, j, str))
        for i in range(self.length):
            for j in range(self.length):
                if self.matrix_w[i][j] != maxsize:
                    self.sum += self.matrix_w[i][j]

    def eh(self, i, j):
        if 4 < self.length < 10:
            return 4.4
        elif 9 < self.length < 15:
            return 5.3
        elif 14 < self.length < 20:
            return 5.8
        elif 19 < self.length < 25:
            return 6.1
        elif 24 < self.length < 30:
            return 6.3
        elif self.length > 29:
            return 6.5
        else:
            return 0

    def es(self, i, j):
        '''Подскажите, как вычислять это (a stacked pair).'''
        return randint(0, 3) - 3

    def eb(self, i, j, a, b):
        '''И это (a bulge or interior loop). То есть это же какие-то табличные данные?'''
        return randint(0, 3) - 4

    def VBI(self, i, j):
        local_minimum = maxsize
        for a in range(i + 1, j - 1):
            for b in range(i + 2, j):
                if a < b and a - i + j - b > 2:
                    if self.eb(i, j, a, b) + self.matrix_v[a][b] < local_minimum:
                        local_minimum = self.eb(i, j, a, b) + self.matrix_v[a][b]
        return local_minimum

    def monte_carlo(self, str, number):
        optimal_stricture = ""
        for i in range(0, len(str), 3):
            optimal_stricture += self.d[str[i:i + 3]][0]
        self._matrix_v(optimal_stricture)
        optimal_energy = self.sum
        for k in range(2, number):
            this_stricture = ""
            for i in range(0, len(str), 3):
                j = len(self.d[str[i:i + 3]])
                this_stricture += self.d[str[i:i + 3]][j - 1]
            self._matrix_v(this_stricture)
            if self.sum < optimal_energy:
                optimal_stricture = this_stricture
                optimal_energy = self.sum

        return optimal_stricture, optimal_energy


if __name__ == '__main__':
    aminoacids = 'PheSerProProAlaUrpUrpCysPhe'
    zuker = MyClass()
    print(zuker.monte_carlo(aminoacids, 20))

File: zuker.scala
import scala.collection.mutable.ArrayBuffer

/**
  * Created by Alexander on 07.02.2016.
  */
class Main {
  var matrix_w = ArrayBuffer[ArrayBuffer[Double]]()
  var matrix_v = ArrayBuffer[ArrayBuffer[Double]]()
  var length = 0
  var sum: Double = 0
  var d: scala.collection.immutable.Map[java.lang.String, Array[java.lang.String]] =
    Map("Phe" -> Array[java.lang.String]("UUA", "UUC", "UUG", "UUU"), "Leu" -> Array[java.lang.String]("CUA", "CUC", "CUG", "CUU"),
      "Ile" -> Array[java.lang.String]("AUA", "AUC", "AUU"), "Met" -> Array[java.lang.String]("AUG"),
      "Ser" -> Array[java.lang.String]("UCA", "UCC", "UCG", "UCU", "AGU", "AGC"), "Pro" -> Array[java.lang.String]("CCA", "CCC", "CCG", "CCU"),
      "Uhr" -> Array[java.lang.String]("ACA", "ACC", "ACG", "ACU"), "Ala" -> Array[java.lang.String]("GCA", "GCC", "GCG", "GCU"),
      "Uyr" -> Array[java.lang.String]("UAC", "UAU"), "His" -> Array[java.lang.String]("CAC", "CAU"), "Gln" -> Array[java.lang.String]("CAA", "CAG"),
      "Asn" -> Array[java.lang.String]("AAC", "AAU"), "Lys" -> Array[java.lang.String]("AAA", "AAG"), "Asp" -> Array[java.lang.String]("GAC", "GAU"),
      "Glu" -> Array[java.lang.String]("GAA", "GAG"), "Cys" -> Array[java.lang.String]("UGC", "UGU"), "Urp" -> Array[java.lang.String]("UGG"),
      "Arg" -> Array[java.lang.String]("CGA", "CGC", "CGG", "CGU", "AGA", "AGG"),
      "Gly" -> Array[java.lang.String]("GGA", "GGC", "GGG", "GGU"))

  def is_complement(x: Char, y: Char): Boolean =
    if (((x == 'A') && (y == 'U')) || ((x == 'U') && (y == 'A')) || ((x == 'G') && (y == 'C')) || ((x == 'C') && (y == 'G'))) true else false


  def _matrix_v(str: java.lang.String) = {
    val a = new Main()
    a.length = str.length()
    var helping_list = ArrayBuffer[Double]()
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        helping_list += 8.4
      }
      a.matrix_v += helping_list
      helping_list = ArrayBuffer[Double]()

    }
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        if ((i > j - 4) && (i < j)) a.matrix_v(i)(j) = 4.4
      }
    }
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        if (i < j) {
          var my_list = ArrayBuffer[Double]()
          my_list += a.eh(i, j)
          my_list += a.es(i, j) + a.matrix_v(i + 1)(j - 1)
          my_list += a.VBI(i, j)
          var local_minimum = my_list(0)
          for (k <- 1 to my_list.length - 1) {
            if (local_minimum > my_list(k)) local_minimum = my_list(k)
          }
          a.matrix_v(i)(j) = local_minimum
        }
      }
    }
    a.zuker(str)
  }

  def eh(i: Integer, j: Integer): Double = {
    val b = new Main()
    if ((b.length > 4) && (b.length < 10)) 4.4
    else if ((b.length > 9) && (b.length < 15)) 5.3
    else if ((b.length > 14) && (b.length < 20)) 5.8
    else if ((b.length > 19) && (b.length < 25)) 6.1
    else if ((b.length > 24) && (b.length < 30)) 6.3
    else if (b.length > 29) 6.5
    else 0
  }

  def es(i: Integer, j: Integer): Double = -0.5

  def eb(i: Integer, j: Integer, a: Integer, b: Integer): Double = -0.5

  def VBI(i: Integer, j: Integer): Double = {
    val c = new Main()
    var local_minimum: Double = 10000000
    for (a <- i + 1 to j - 2) {
      for (b <- i + 2 to j - 1) {
        if ((a < b) && (a - i + j - b > 2)) {
          val d = c.eb(i, j, a, b) + c.matrix_v(a)(b)
          if (local_minimum > d) local_minimum = d
        }
      }
    }
    local_minimum
  }

  def one(i: Integer, j: Integer, str: java.lang.String): Double = {
    val a = new Main()
    var flag: Boolean = false
    for (n <- i + 1 to j) {
      if (a.is_complement(str(i), str(j))) flag = true
    }
    if (flag) 10000000
    else a.matrix_w(i + 1)(j)
  }

  def two(i: Integer, j: Integer, str: java.lang.String): Double = {
    val a = new Main()
    var flag: Boolean = false
    for (n <- i + 0 to j - 1) {
      if (a.is_complement(str(j), str(n))) flag = true
    }
    if (flag) 10000000
    else a.matrix_w(i)(j - 1)
  }

  def three(i: Integer, j: Integer, str: java.lang.String): Double = {
    val a = new Main()
    if (a.is_complement(str(i), str(j))) 10000000
    var local_minimum: Double = a.matrix_w(i)(i + 1) + a.matrix_w(i + 2)(j)
    for (k <- i + 2 to j - 2) {
      if (a.matrix_w(i)(k) + a.matrix_w(k + 1)(j) < local_minimum) local_minimum = a.matrix_w(i)(k) + a.matrix_w(k + 1)(j)
    }
    local_minimum
  }

  def four(i: Integer, j: Integer, str: java.lang.String): Double = {
    val a = new Main()
    if (!a.is_complement(str(i), str(j))) 10000000
    else a.matrix_v(i)(j)
  }

  def zuker(str: java.lang.String) = {
    val a = new Main()
    var helping_list = ArrayBuffer[Double]()
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        helping_list += 0
      }
      a.matrix_w += helping_list
      helping_list = ArrayBuffer[Double]()
    }
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        if ((i > j - 4) && (i < j)) a.matrix_w(i)(j) = 10000000
      }
    }
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        if ((i < j) && (j - i > 4)) {
          var my_list = ArrayBuffer[Double]()
          my_list += a.one(i, j, str)
          my_list += a.two(i, j, str)
          my_list += a.three(i, j, str)
          my_list += a.four(i, j, str)
          var local_minimum = my_list(0)
          for (k <- 1 to my_list.length - 1) {
            if (local_minimum > my_list(k)) local_minimum = my_list(k)
          }
          a.matrix_w(i)(j) = local_minimum
        }
      }
    }
    for (i <- 0 to a.length - 1) {
      for (j <- 0 to a.length - 1) {
        if (a.matrix_w(i)(j) < 10000000) a.sum = a.matrix_w(i)(j)
      }
    }
  }

  def count(str: java.lang.String, number: Integer): Double = {
    val a = new Main()
    var optimal_structure: java.lang.String = ""
    var s: java.lang.String = ""
    for (i <- 0 to a.length - 1) {
      if (i % 3 != 0) {
        s += str(i)
      }
      else if (i % 3 == 0) {
        s += str(i)
        val t: Array[java.lang.String] = a.d(s)
        optimal_structure += t(0)
        s = ""
      }
    }
    s = ""
    a._matrix_v(optimal_structure)
    var optimal_energy = a.sum
    for (k <- 1 to number) {
      var this_structure = ""
      for (i <- 0 to a.length - 1) {
        if (i % 3 != 0) {
          s += str(i)
        }
        else if (i % 3 == 0) {
          s += str(i)
          val t: Array[java.lang.String] = a.d(s)
          val j = t.length - 1 //Здесь можно выбирать случайное число из отрезка [0, length - 1]
          this_structure += t(j)
          s = ""
        }
      }
      a._matrix_v(this_structure)
      if (a.sum < optimal_energy) {
        optimal_energy = a.sum
        optimal_structure = this_structure
      }
    }
    print(optimal_structure)
    optimal_energy
  }

  def main(args: Array[java.lang.String]): Unit = {
    val k = new Main()
    print(k.count("ProGlu", 20))
  }
}
